File: mobx/src/test/java/com/ivansadovyi/mobx/AutorunTest.java
package com.ivansadovyi.mobx;

import org.junit.Test;

import static com.ivansadovyi.mobx.MobxOperators.autorun;
import static org.mockito.Mockito.*;

public class AutorunTest {

	@Test
	public void testSingleAutorun() {
		final TrackedObservable<Integer> observableValue = new TrackedObservable<>(0);
		Runnable autorunCallback = spy(new Runnable() {
			@Override
			public void run() {
				observableValue.get();
			}
		});
		autorun(autorunCallback);

		verify(autorunCallback, times(1)).run();
		clearInvocations(autorunCallback);

		observableValue.set(1);

		verify(autorunCallback, times(1)).run();
	}

	@Test
	public void testMultipleAutoruns() {
		final TrackedObservable<Integer> observableValue = new TrackedObservable<>(0);
		Runnable autorunCallback1 = spy(new Runnable() {
			@Override
			public void run() {
				observableValue.get();
			}
		});
		Runnable autorunCallback2 = spy(new Runnable() {
			@Override
			public void run() {
				observableValue.get();
			}
		});
		autorun(autorunCallback1);
		autorun(autorunCallback2);

		verify(autorunCallback1, times(1)).run();
		verify(autorunCallback2, times(1)).run();
		clearInvocations(autorunCallback1, autorunCallback2);

		observableValue.set(1);

		verify(autorunCallback1, times(1)).run();
		verify(autorunCallback2, times(1)).run();
	}

	@Test
	public void testDispose() {
		final TrackedObservable<Integer> observableValue = new TrackedObservable<>(0);
		Runnable autorunCallback = spy(new Runnable() {
			@Override
			public void run() {
				observableValue.get();
			}
		});
		Disposable disposable = autorun(autorunCallback);
		clearInvocations(autorunCallback);

		disposable.dispose();
		observableValue.set(1);

		verify(autorunCallback, times(0)).run();
	}
}

File: mobx/src/test/java/com/ivansadovyi/mobx/AutorunTest.kt
package com.ivansadovyi.mobx

import com.nhaarman.mockitokotlin2.clearInvocations
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.verifyZeroInteractions
import org.junit.Test
import org.mockito.Mockito.spy

class AutorunTest {

	@Test
	fun testAutorunShouldBeCalledImmediately() {
		val autorunBody = spy { }
		autorun(autorunBody)

		verify(autorunBody).invoke()
	}

	@Test
	fun testAutorunShouldBeCalledImmediatelyAndOnObservableChange() {
		val observable = TrackedObservable(0)
		val autorunBody = spy<AutorunBody> { observable.value }
		autorun(autorunBody)

		verify(autorunBody).invoke()
		clearInvocations(autorunBody)

		observable.value = 1

		verify(autorunBody).invoke()
	}

	@Test
	fun testAutorunShouldNotBeCalledAfterDispose() {
		val observable = TrackedObservable(0)
		val autorunBody = spy<AutorunBody> { observable.value }
		val autorun = autorun(autorunBody)

		verify(autorunBody).invoke()
		clearInvocations(autorunBody)

		observable.value++

		verify(autorunBody).invoke()
		clearInvocations(autorunBody)

		autorun.dispose()
		observable.value++

		verifyZeroInteractions(autorunBody)
	}
}
